File: Sources/XMTP/Codecs/Composite.swift
//
//  CompositeCodec.swift
//
//
//  Created by Pat Nakajima on 12/22/22.
//

import XMTPProto

typealias Composite = Xmtp_MessageContents_Composite

let ContentTypeComposite = ContentTypeID(authorityID: "xmtp.org", typeID: "composite", versionMajor: 1, versionMinor: 0)

extension Composite.Part {
	init(encodedContent: EncodedContent) {
		self.init()
		element = .part(encodedContent)
	}

	init(composite: Composite) {
		self.init()
		element = .composite(composite)
	}
}

struct CompositeCodec: ContentCodec {
	public typealias T = DecodedComposite

	public var contentType: ContentTypeID {
		ContentTypeComposite
	}

	public func encode(content: DecodedComposite) throws -> EncodedContent {
		let composite = toComposite(content: content)
		var encoded = EncodedContent()
		encoded.type = ContentTypeComposite
		encoded.content = try composite.serializedData()
		return encoded
	}

	public func decode(content encoded: EncodedContent) throws -> DecodedComposite {
		let composite = try Composite(serializedData: encoded.content)
		let decodedComposite = fromComposite(composite: composite)
		return decodedComposite
	}

	func toComposite(content decodedComposite: DecodedComposite) -> Composite {
		var composite = Composite()

		if let content = decodedComposite.encodedContent {
			composite.parts = [Composite.Part(encodedContent: content)]
			return composite
		}

		for part in decodedComposite.parts {
			if let encodedContent = part.encodedContent {
				composite.parts.append(Composite.Part(encodedContent: encodedContent))
			} else {
				composite.parts.append(Composite.Part(composite: toComposite(content: part)))
			}
		}

		return composite
	}

	func fromComposite(composite: Composite) -> DecodedComposite {
		var decodedComposite = DecodedComposite()

		if composite.parts.count == 1, case let .part(content) = composite.parts.first?.element {
			decodedComposite.encodedContent = content
			return decodedComposite
		}

		decodedComposite.parts = composite.parts.map { fromCompositePart(part: $0) }

		return decodedComposite
	}

	func fromCompositePart(part: Composite.Part) -> DecodedComposite {
		var decodedComposite = DecodedComposite()

		switch part.element {
		case let .part(encodedContent):
			decodedComposite.encodedContent = encodedContent
		case let .composite(composite):
			decodedComposite.parts = composite.parts.map { fromCompositePart(part: $0) }
		case .none:
			return decodedComposite
		}

		return decodedComposite
	}
}

File: library/src/main/java/org/xmtp/android/library/codecs/Composite.kt
package org.xmtp.android.library.codecs

import org.xmtp.proto.message.contents.CompositeOuterClass
import org.xmtp.proto.message.contents.CompositeOuterClass.Composite.Part

typealias Composite = org.xmtp.proto.message.contents.CompositeOuterClass.Composite

val ContentTypeComposite = ContentTypeIdBuilder.builderFromAuthorityId(
    authorityId = "xmtp.org",
    typeId = "composite",
    versionMajor = 1,
    versionMinor = 0
)

class CompositePartBuilder {
    companion object {
        fun buildFromEncodedContent(encodedContent: EncodedContent): CompositeOuterClass.Composite.Part {
            return CompositeOuterClass.Composite.Part.newBuilder().also {
                it.part = encodedContent
            }.build()
        }

        fun buildFromComosite(composite: Composite): CompositeOuterClass.Composite.Part {
            return CompositeOuterClass.Composite.Part.newBuilder().also {
                it.composite = composite
            }.build()
        }
    }
}

@Deprecated(
    message = "This content type will be removed in future versions (see here for more info -> https://community.xmtp.org/t/xip-19-deprecate-the-composite-codec/525). Consider using a custom content type (https://github.com/xmtp/xmtp-android?tab=readme-ov-file#handle-custom-content-types) instead"
)
class CompositeCodec : ContentCodec<DecodedComposite> {
    override val contentType: ContentTypeId
        get() = ContentTypeComposite

    override fun encode(content: DecodedComposite): EncodedContent {
        val composite = toComposite(content)
        return EncodedContent.newBuilder().also {
            it.type = ContentTypeComposite
            it.content = composite.toByteString()
        }.build()
    }

    override fun decode(content: EncodedContent): DecodedComposite {
        val composite = Composite.parseFrom(content.content)
        return fromComposite(composite = composite)
    }

    override fun fallback(content: DecodedComposite): String? {
        return null
    }

    private fun toComposite(decodedComposite: DecodedComposite): Composite {
        return Composite.newBuilder().also {
            val content = decodedComposite.encodedContent
            if (content != null) {
                it.addParts(CompositePartBuilder.buildFromEncodedContent(content))
                return it.build()
            }
            for (part in decodedComposite.parts) {
                val encodedContent = part.encodedContent
                if (encodedContent != null) {
                    it.addParts((CompositePartBuilder.buildFromEncodedContent(encodedContent)))
                } else {
                    it.addParts((CompositePartBuilder.buildFromComosite(toComposite(part))))
                }
            }
        }.build()
    }

    private fun fromComposite(composite: Composite): DecodedComposite {
        val decodedComposite = DecodedComposite()

        if (composite.partsList.size == 1 && composite.partsList.first().elementCase == Part.ElementCase.PART) {
            decodedComposite.encodedContent = composite.partsList.first().part
            return decodedComposite
        }
        decodedComposite.parts = composite.partsList.map { fromCompositePart(part = it) }
        return decodedComposite
    }

    private fun fromCompositePart(part: Part): DecodedComposite {
        return when (part.elementCase) {
            Part.ElementCase.PART -> {
                DecodedComposite(emptyList(), part.part)
            }

            Part.ElementCase.COMPOSITE -> {
                DecodedComposite(part.composite.partsList.map { fromCompositePart(it) })
            }

            else -> DecodedComposite()
        }
    }
}
